src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -2,89 +2,147 @@
 
-import java.util.ArrayList;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 
+import static com.epam.aidial.core.storage.BlobStorageUtil.PATH_SEPARATOR;
+
+@Slf4j
+@RequiredArgsConstructor
 public class TokenStatsTracker {
-    private final Map<String, TraceContext> traceIdToContext = new ConcurrentHashMap<>();
+    public static final String DEPLOYMENT_COST_STATS_BUCKET = "deployment_cost_stats";
+    public static final String DEPLOYMENT_COST_STATS_LOCATION = DEPLOYMENT_COST_STATS_BUCKET + PATH_SEPARATOR;
+
+    private final Vertx vertx;
+    private final ResourceService resourceService;
 
-    public void startSpan(ProxyContext context) {
-        TraceContext traceContext = traceIdToContext.computeIfAbsent(context.getTraceId(), k -> new TraceContext());
-        traceContext.addSpan(context);
+    /**
+     * Starts current span.
+     * <p>
+     *     Note. The method is blocking and shouldn't be run in the event loop thread.
+     * </p>
+     */
+    public Future<Void> startSpan(ProxyContext context) {
+        return vertx.executeBlocking(() -> {
+            ResourceDescription resource = toResource(context.getTraceId());
+            resourceService.computeResource(resource, json -> {
+                TraceContext traceContext = ProxyUtil.convertToObject(json, TraceContext.class);
+                if (traceContext == null) {
+                    traceContext = new TraceContext();
+                }
+                traceContext.addSpan(context);
+                return ProxyUtil.convertToString(traceContext);
+            });
+            return null;
+        }, false);
     }
 
     public Future<TokenUsage> getTokenStats(ProxyContext context) {
-        TraceContext traceContext = traceIdToContext.get(context.getTraceId());
-        if (traceContext == null) {
-            return Future.succeededFuture();
-        }
-        return traceContext.getStats(context);
+        return vertx.executeBlocking(() -> {
+            ResourceDescription resource = toResource(context.getTraceId());
+            String json = resourceService.getResource(resource);
+            TraceContext traceContext = ProxyUtil.convertToObject(json, TraceContext.class);
+            if (traceContext == null) {
+                return null;
+            }
+            return traceContext.getStats(context);
+        }, false);
     }
 
-    public void endSpan(ProxyContext context) {
+    /**
+     * Ends current span.
+     */
+    public Future<Void> endSpan(ProxyContext context) {
         ApiKeyData apiKeyData = context.getApiKeyData();
         if (apiKeyData.getPerRequestKey() == null) {
-            traceIdToContext.remove(context.getTraceId());
+            return vertx.executeBlocking(() -> {
+                ResourceDescription resource = toResource(context.getTraceId());
+                resourceService.deleteResource(resource, EtagHeader.ANY);
```

Comment:
Does it work for this case:

Application uses two models?

Reply:
yes, we delete trace for root requests only, e.g. chat back end -> app. Core knows that request doesn't have any per request API key with that request so we are safe to delete. Otherwise app may make calls to two model A and B and we have to wait for the request to the app is completed and then delete the trace.